pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/extended/MetadataStoreExtended.java

```diff
@@ -72,8 +72,23 @@ CompletableFuture<Stat> put(String path, byte[] value, Optional<Long> expectedVe
      *
      * @param listener
      *            the session listener
+     * @return a runnable that can be used to unregister the listener
      */
-    void registerSessionListener(Consumer<SessionEvent> listener);
+    Runnable registerCancellableSessionListener(Consumer<SessionEvent> listener);
```

Comment:
can we keep the name as-is? `registerSessionListener`?

Reply:
That would break binary compatibility for Metadata Store API consumers such as plugins. There's a default method with the old name.
There are solutions for changing the return type while keeping binary compatibility, but that would require a compiler plugin. Some solutions are explained in this article: https://www.morling.dev/blog/refining-return-type-java-methods-without-breaking-backwards-compatibility/ and in the comments.

Reply:
oh sorry, I made a mistake and called it `addSessionListener`, I'll fix that.

Reply:
Fixed it.

Reply:
Can the parent class just remove all listeners when closed?

Reply:
> Can the parent class just remove all listeners when closed?

@heesung-sn What are you referring as the "parent class"? We don't currently have any dependency injection or application eventing solution in Pulsar Java code so creating such parent classes could easily make things much more complicated than what is done in this PR. Each class will keep track of what listeners it has registered and execute the callbacks when the instance is closed.

The metadata store cannot be the "parent class" since it doesn't know when an individual pulsar broker is stopped. A metadata store can be shared and used across multiple pulsar broker instances in tests. That's why this solution is needed in the first place.

Reply:
I was referring to `AbstractMetadataStore`. Can AbstractMetadataStore.close() remove the listeners?